pingcap/tidb-operator
# Remove failed pod and pvcs of PD, TIKV and TiFlash components for auto failure recovery

### What problem does this PR solve?
Handle unplanned failures of K8s nodes hosting stateful Pods of a Tidb cluster during failover.

Failed stateful Pods are not removed after the K8s node goes down or has become unserviceable. Similarly, in case of local attached storage, the PVCs used by the failed Pods are not removed. After a hardware failure, the kubelet of the K8s node would not respond hence the Pods running on it cannot be deleted gracefully. In such scenario, the Pods have to be forcefully deleted. In case of TiKV/TiFlash failure recovery, the failure store also needs to be deleted and be in Tombstone state before the PVCs can be removed.

The auto failure recovery is handled for the following scenarios -
1. K8s node with network mounted storage like EBS fails
2. Disk of a k8s node with local attached storage becomes read-only
3. K8s node with local attached storage goes down

### What is changed and how does it work?
TiDB operator supports automatic failover by scaling Pods automatically. During failover, the operator adds a FailureMember(PD) or FailureStore(TiKV/TiFlash). The FailureMember or FailureStore is not removed until the failed member or store has been removed. With this changes, the recovery of the failed Pod is attempted first before removing the failed member or store after FailureMember or FailureStore have been added. The auto recovery has two parts - detection and failure recovery. During detection, the TiDB operator checks the Pod and the Node conditions and the Pod Phase to determine if the underlying k8s node has become unavailable and then sets a HostDown flag in the FailureMember or FailureStore.

During failure recovery, if the member is unhealthy or store is down and corresponding HostDown is set, the TiDB operator attempts a force restart of the Pod. In case of K8s nodes having network mounted storage like EBS, it is expected that the Pod will get re-created on a new compute node and the failed member or store may recover when the Pod is up. In case of local attached storage, a new Pod will be created but will be in unscheduled state as the PVC created on the failure node will be existing. For PD, if the Pod continues to remain unscheduled then the member is deleted and the PVC is removed along with the Pod. For TiKV, if the Pod continues to remain unscheduled then the store is deleted and the PVC will be removed along with the Pod after the store becomes Tombstone. MemberDeleted or StoreDeleted is then set in FailureMember or FailureStore respectively. When the unscheduled Pod and the old PVC are removed, K8s will create a new Pod and PVC.

The detection is controlled with an operator flag `detectNodeFailure` which is added in CLIConfig. The failure recovery is controlled at an individual Tidb cluster level with an `autoFailureRecovery` flag. There is also a `podHardRecoveryPeriod` variable in operator which is added in CLIConfig and it indicates the waiting time to enforce a failure recovery by setting the HostDown in FailureStore if the store remains down for a long time and the operator cannot detect node failure.

### Code changes


### Tests
  - Run test suite with flags disabled
  - Modify charts/tidb-operator/templates/controller-manager-deployment.yaml to enable detectNodeFailure and run the test suite
  - Modify tests/e2e/tidbcluster/tidbcluster.go to enable autoFailureRecovery as well and run the test suite
  - Deploy operator with the following configuration cli configuration`-detect-node-failure=true` and `-pod-hard-recovery-period=1h`
  - Create a TidbCluster. Add a annotation in the metadata `kubernetes.io/auto-failure-recovery: "true"`
  - Simulate node failure (one way to do this is to stop the host machine) on which a TiKV pod is running and the node has local attached storage
  - Observe the Store become Down and after tikvFailoverPeriod an extra TiKV is added and the HostDown is set and the failed pod is restarted.
  - Verify that after 10 minutes the Store is deleted. When it becomes Tombstone the restarted Pod and old PVC gets removed and StoreDeleted is set in FailureStore. As the Pod and PVC are removed, a new Pod and PVC are created on a different k8s node.
  - Repeat for PD and TiFlash as well
  - Repeat the test with K8s nodes having openEBS volumes.
    - Verify that after the pod restart, a new pod is able to come up and the member or store become healthy again and the failure recovery does not proceed further. Verify that the extra pod is created in parallel as part of auto failover.
  - Repeat the test for read-only disk scenario with K8s nodes having local attached storage.
    - Make a disk read-only by triggering sysrq as mentioned in https://www.kernel.org/doc/html/v4.10/admin-guide/sysrq.html or by setting setro option for the block device as mentioned in https://man7.org/linux/man-pages/man8/blockdev.8.html
    - Verify that the Member becomes unhealthy or Tikv/Tiflash Store become Disconnected and then Down.
    - After failover period verify that the FailureMember or FailureStore is added. Verify that the HostDown in FailureMember or FailureStore gets set and the failure recovery happens similar to the node down case. Verify that the extra pod is created in parallel as part of auto failover.

### Side effects

- [ ] Other side effects: NA

### Related changes


### Release Notes
Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note

```
